subsys/net/ip/dhcpv6.c

```diff
@@ -0,0 +1,2205 @@
+		}
+
+		return iface->config.dhcpv6.timeout;
+	case NET_DHCPV6_RENEWING:
+		if (now >= iface->config.dhcpv6.t2) {
+			dhcpv6_enter_state(iface, NET_DHCPV6_REBINDING);
+			return iface->config.dhcpv6.timeout;
+		}
+
+		iface->config.dhcpv6.retransmissions++;
+		iface->config.dhcpv6.retransmit_timeout =
+			dhcpv6_next_retransmit_time(
+				iface->config.dhcpv6.retransmit_timeout,
+				DHCPV6_REN_MAX_RT);
+
+		(void)dhcpv6_send_renew(iface);
+		dhcpv6_set_timeout(iface, iface->config.dhcpv6.retransmit_timeout);
+
+		if (iface->config.dhcpv6.timeout > iface->config.dhcpv6.t2) {
+			iface->config.dhcpv6.timeout = iface->config.dhcpv6.t2;
+		}
+
+		return iface->config.dhcpv6.timeout;
+	case NET_DHCPV6_REBINDING:
+		if (now >= iface->config.dhcpv6.expire) {
+			dhcpv6_enter_state(iface, NET_DHCPV6_SOLICITING);
+			return iface->config.dhcpv6.timeout;
+		}
+
+		iface->config.dhcpv6.retransmissions++;
+		iface->config.dhcpv6.retransmit_timeout =
+			dhcpv6_next_retransmit_time(
+				iface->config.dhcpv6.retransmit_timeout,
+				DHCPV6_REB_MAX_RT);
+
+		(void)dhcpv6_send_rebind(iface);
+		dhcpv6_set_timeout(iface, iface->config.dhcpv6.retransmit_timeout);
+
+		if (iface->config.dhcpv6.timeout > iface->config.dhcpv6.expire) {
+			iface->config.dhcpv6.timeout = iface->config.dhcpv6.expire;
+		}
+
+		return iface->config.dhcpv6.timeout;
+	case NET_DHCPV6_INFO_REQUESTING:
+		break;
+	case NET_DHCPV6_BOUND:
+		dhcpv6_enter_state(iface, NET_DHCPV6_RENEWING);
+		return iface->config.dhcpv6.timeout;
+	}
+
+	return UINT64_MAX;
+}
+
+static void dhcpv6_timeout(struct k_work *work)
+{
+	uint64_t timeout_update = UINT64_MAX;
+	int64_t now = k_uptime_get();
+	struct net_if_dhcpv6 *current, *next;
+
+	ARG_UNUSED(work);
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&dhcpv6_ifaces, current, next, node) {
+		struct net_if *iface = CONTAINER_OF(
+			CONTAINER_OF(current, struct net_if_config, dhcpv6),
+			struct net_if, config);
+		uint64_t next_timeout;
+
+		next_timeout = dhcpv6_manage_timers(iface, now);
+		if (next_timeout < timeout_update) {
+			timeout_update = next_timeout;
+		}
+	}
+
+	k_mutex_unlock(&lock);
+
+	if (timeout_update != UINT64_MAX) {
+		if (now > timeout_update) {
+			timeout_update = 0ULL;
+		} else {
+			timeout_update -= now;
+		}
+
+		NET_DBG("Waiting for %llums", timeout_update);
+		k_work_reschedule(&dhcpv6_timeout_work, K_MSEC(timeout_update));
+	}
+}
+
+static void dhcpv6_iface_event_handler(struct net_mgmt_event_callback *cb,
+				       uint32_t mgmt_event, struct net_if *iface)
+{
+	sys_snode_t *node = NULL;
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	SYS_SLIST_FOR_EACH_NODE(&dhcpv6_ifaces, node) {
+		if (node == &iface->config.dhcpv6.node) {
+			break;
+		}
+	}
+
+	if (node == NULL) {
+		goto out;
+	}
+
+	if (mgmt_event == NET_EVENT_IF_DOWN) {
+		NET_DBG("Interface %p going down", iface);
+		dhcpv6_set_timeout(iface, UINT64_MAX);
+		dhcpv6_reschedule();
```

Comment:
As both branches call reschedule, we could move the function call outside of the if statements.